src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs

```diff
@@ -2,10 +2,17 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Buffers;
+
 namespace System.Text.Json
 {
     public static partial class JsonSerializer
     {
+        // The maximum length of a byte array before we ask for the actual transcoded byte size.
+        // The "int.MaxValue / 2" is used because max byte[] length is a bit less than int.MaxValue
+        // and because we don't want to allocate such a large buffer if not necessary.
+        private const int MaxArrayLengthBeforeCalculatingSize = int.MaxValue / 2 / JsonConstants.MaxExpansionFactorWhileTranscoding;
```

Comment:
Why are we dividing by 2 first?

Reply:
The reason for dividing by 2 is explained. Are you suggesting it is after the / 3? (same result of /6).

We could divide by 4 or some other number instead of 2. It's a trade-off: asking for the transcoded size is CPU expensive, but allocating unnecessary memory can be wasteful (but not so much if coming from a pool).

Reply:
I was still confused by why we divide by 2. Re-reading the comment is starting to make some sense, but let me verify:
`The "int.MaxValue / 2" is used because max byte[] length is a bit less than int.MaxValue`

> We could divide by 4 or some other number instead of 2. It's a trade-off: asking for the transcoded size is CPU expensive, but allocating unnecessary memory can be wasteful (but not so much if coming from a pool).

Ah yes, now that makes sense. So, the division by 2 is just an arbitrary value then, correct? So, rather than say 2 billion, we chose 1 billion because we might as well get the actual byte count once we are at above 1 billion. It could have been 500 million (or 250 million), as well.

In that case, maybe we should set the threshold of 1 million since after that the array pool allocates an array anyway so its probably not worth it going above that. In fact, maybe this change should only apply for lengths < 1 million, and for anything above that we keep using the old code-path.